### DISCUSSION TITLE
implement NonEmptyList.tryZip

### DISCUSSION THREAD

**Discussion by robkuz** [gmpl/FSharpPlus#292] [Ideas]
the zip implementation uses `List.zip` underneath which will fail with a runtime exception

we should have a save zip version

    let tryZip (xs: NonEmptyList<'x>) (ys: NonEmptyList<'y>) =
        if xs.Length <> ys.Length then
            None
        else
            NonEmptyList.zip xs ys |> Some

**Comment by gusty** [comment]
Yes, it fails when list lenghts are different.

I personally don't like this behavior, but that seems to be the standard for List and Arrays in F#, whereas for Seq they use the shortest which is the functionality I would expect otherwise.

So, another option to have a safe version is to define a `zipShortest` function.

What do you think?

**Comment by robkuz** [comment]
well, all this comes down to subjective prefs.

What I do these days is this
On types fully under my control I define a
- `tryXXX`
- `unsafeXXX`

for types not under my control
i check if they follow the above pattern but do not implement both functions.
if either is missing I might implement it or even alias the existing one.

coming back to your question: no I would not implement a `zipShortest` at least not primeraly.
I would (and I have) implemented `tryZip:: NonEmptyList<'a> -> NonEmptyList<'b> -> Option<NonEmptyList<'a * 'b>>`
however to follow the above pattern.

**Comment by gusty** [comment]
Ok. I don’t see any reason why not to implement your suggestion. 

Do you want to submit a PR ?

**Comment by robkuz** [comment]
will do

**Comment by 3Rafal** [comment]
I've been inspired by this suggestion, and thought that it could be a good idea to add safe zip to F#+. I've implemented it for FSharp.Core types that have unsafe zip implementation and created a PR https://github.com/fsprojects/FSharpPlus/pull/370 . I wonder if it could be a first step, then I could expand it to all types with unsafe zip in F#+

**Comment by gusty** [comment]
> the zip implementation uses `List.zip` underneath which will fail with a runtime exception
> 

Maybe we should change that, make our generic `zip` safe.

**Comment by 3Rafal** [comment]
@gusty , I guess this can be closed

**Comment by gusty** [comment]
Not sure, let's ask @robkuz if he's happy with your implementation. I think it's not exactly what he was after.
